### DISCUSSION TITLE
Event preroll

### DISCUSSION THREAD

**Discussion by DerekL0576** [tsightler/ring-mqtt#348] [General Discussion]
I've added this under assistance because it's not really a bug and I'm not sure if it's something that can even be added as a new feature.

My Ring doorbell records 8 seconds before a motion event and this is inserted at the start of the event videos on the ring account.  The problem is, I think they do this on the fly when viewing or downloading the video on the Ring site or app.

When I try to record the live feed in HA on a motion event, the start of the video is missing a reference key frame so it's garbled for around the first 2 seconds.  Obviously the video only starts from the motion event and is missing the 8 seconds preroll video.

When viewing the latest event feed in HA, the video is exactly the same as my local recording.  Garbled at the start and missing the 8 seconds preroll.

Could this preroll video ever be accessed by ring-mqtt and inserted at the start of the event stream like Ring does?

**Comment by DerekL0576** [comment]
I just wanted to add that manually recorded live stream video is perfect, it's only garbled at the start if I start recording on a motion event.

**Comment by tsightler** [comment]
I'm going to convert this to a discussion as this is not really an issue.

**Comment by tsightler** [comment]
You can potentially decrease the initial garbling for live streams by increasing the UDP buffers on your Linux setup (rmem_max/rmem_default) as one of the issues is that Ring sends a huge burst of traffic super fast and overruns the buffers before the local RTP client has the chance to get them.  It won't be perfect even then, but it might help.  Note that the current code does not modify the encoded stream in any way, it's literally the exact data from Ring servers just streamed through a pipeline.  That being said, since you say you see exactly this behavior for recorded streams, which are literally just replaying the stream directly from the recording file stored on Ring's cloud, I'm not sure that it would help.

Because of this, it's never going to be better than what you get when you download the Ring recoding, although it should be basically the same.  With the dev version the recording my start a little bit faster since it's using push notifications instead of polling which means it should get the motion event more quickly.

Now, when you talk about pre-roll, unfortunately I don't have any cameras that support this feature so I'm in the dark about it somewhat, however, based on what I've read about it, the cameras simply keep a small buffer of recording which is completely separate from the actual event recording.  For battery cameras it's a 4 second black & white recording, while for wired cameras it's a 6 second full color recording.  Based on what I can see, the pre-roll recording is completely separate, it just basically mushes them together in the timeline so that they play back-to-back.

I assume it's probably possible to access it like any other recording, but as I have no cameras that support it, and there is no existing abstraction in ring-client-api, there's not much I'll be able to do about adding it for now, unless someone is willing to give me access to their cameras to poke around at the API and see what I can discover, or someone wants to submit a PR for it.
